js/models/conversations.js

```diff
@@ -169,6 +185,95 @@
             return textsecure.messaging.syncVerification(number, state, key);
         });
     },
+    getIdentityKeys: function() {
+        var lookup = {};
+
+        if (this.isPrivate()) {
+            return textsecure.storage.protocol.loadIdentityKey(this.id).then(function(key) {
+                lookup[this.id] = key;
+                return lookup;
+            }.bind(this)).catch(function(error) {
+                console.log(
+                    'getIdentityKeys error for conversation',
+                    this.id,
+                    error && error.stack ? error.stack : error
+                );
+                return lookup;
+            }.bind(this));
+        } else {
+            return Promise.all(this.contactCollection.map(function(contact) {
+                return textsecure.storage.protocol.loadIdentityKey(contact.id).then(function(key) {
+                    lookup[contact.id] = key;
+                }).catch(function(error) {
+                    console.log(
+                        'getIdentityKeys error for group member',
+                        contact.id,
+                        error && error.stack ? error.stack : error
+                    );
+                });
+            })).then(function() {
+                return lookup;
+            });
+        }
+    },
+    replay: function(error) {
+        var replayable = new textsecure.ReplayableError(error);
+        return replayable.replay().catch(function(error) {
+            console.log(
+                'decryptOldIncomingKeyErrors error:',
+                error && error.stack ? error.stack : error
+            );
+        });
+    },
+    decryptOldIncomingKeyErrors: function() {
+        if (this.get('decryptedOldIncomingKeyErrors')) {
+            return Promise.resolve();
+        }
+
+        var messages = this.messageCollection.filter(function(message) {
+            var errors = message.get('errors');
+            if (!errors || !errors[0]) {
+                return false;
+            }
+            var error = errors[0];
+            if (error.name !== 'IncomingIdentityKeyError') {
```

Comment:
I'm not familiar with the data model, but it seems like we'd want to see if *any* of the errors was an IncomingIdentityKeyError. Or do we somehow *know* that it would be the first error in 1:1 and group convos?